src/NUnitEngine/nunit.engine/Services/ExtensionService.cs

```diff
@@ -0,0 +1,154 @@
+{
+    public class ExtensionService : Service
+    {
+        private List<ExtensionPoint> _extensionPoints = new List<ExtensionPoint>();
+        private Dictionary<string, ExtensionPoint> _pathIndex = new Dictionary<string, ExtensionPoint>();
+        private Dictionary<Type, ExtensionPoint> _typeIndex = new Dictionary<Type, ExtensionPoint>();
+
+        private List<ExtensionNode> _extensions = new List<ExtensionNode>();
+
+        #region Public Properties
+
+        public IEnumerable<ExtensionPoint> ExtensionPoints
+        {
+            get { return _extensionPoints; }
+        }
+
+        public IEnumerable<ExtensionNode> Extensions
+        {
+            get { return _extensions;  }
+        }
+
+        #endregion
+
+        public ExtensionPoint GetExtensionPoint(string path)
+        {
+            return _pathIndex[path];
+        }
+
+        public ExtensionPoint GetExtensionPoint(Type type)
+        {
+            return _typeIndex[type];
+        }
+
+        public T[] GetExtensions<T>()
+        {
+            var extensions = new List<T>();
+
+            var ep = GetExtensionPoint(typeof(T));
+            foreach (var node in ep.Extensions)
+                extensions.Add((T)node.ExtensionObject);
+
+            return extensions.ToArray();
+        }
+
+        #region Service Overrides
+
+        public override void StartService()
+        {
+            var thisAssembly = Assembly.GetExecutingAssembly();
+            var startDir = new DirectoryInfo(AssemblyHelper.GetDirectoryName(thisAssembly));
+
+            FindExtensionPoints(thisAssembly);
+            FindExtensionsInDirectory(startDir);
+
+            base.StartService();
+        }
+
+        private void FindExtensionPoints(Assembly assembly)
+        {
+            foreach (ExtensionPointAttribute attr in assembly.GetCustomAttributes(typeof(ExtensionPointAttribute), false))
+            {
+                var ep = new ExtensionPoint(attr.Path, attr.Type)
+                {
+                    Description = attr.Description
+                };
+
+                _extensionPoints.Add(ep);
+
+                // TODO: These Adds will throw if the path or type is a duplicate.
+                // We should give an error message instead.
+                _pathIndex.Add(ep.Path, ep);
+                _typeIndex.Add(ep.Type, ep);
+            }
+        }
+
+        private void FindExtensionsInDirectory(DirectoryInfo startDir)
+        {
+            var addinsFiles = startDir.GetFiles("*.addins");
+            if (addinsFiles.Length > 0)
+                foreach (var file in addinsFiles)
+                    ProcessAddinsFile(startDir.FullName, file.FullName);
+            else
+                foreach (var file in startDir.GetFiles("*.dll"))
+                    FindExtensionsInAssembly(file.FullName);
+        }
+
+        private void ProcessAddinsFile(string baseDir, string fileName)
+        {
+            var doc = new XmlDocument();
```

Comment:
I've experimented with this, and I think I prefer to stick with XML, because

1. It takes 2 to 3 times the code to parse even the simplified flat file if you check for errors.

2. It creates another format to deal with, albeit simple.

3. We already use XML for just about everything else: representing tests and results, filters, progress reports, NUnit settings, format of nunit and VS projects.

4. Anyone familiar with Mono.Addins already creates files like this in XML.